File: Algorithm/Programers_연속 펄스 부분 수열의 합.cpp
#include <string>
#include <vector>

using namespace std;

long long solution(vector<int> sequence) {
	auto max = [](long long a, long long b)->long long {return a > b ? a : b; };
	auto max3 = [](long long a, long long b, long long c)->long long { long long temp = a > b ? a : b; return temp > c ? temp : c; };
	vector<long long> upPulse(sequence.size(), 0);
	vector<long long> downPulse(sequence.size(), 0);

	downPulse[0] = -sequence[0];
	upPulse[0] = sequence[0];

	long long answer = max(downPulse[0], upPulse[0]);

	for (long long i = 1; i < sequence.size(); i++) {
		if (i & 1) {
			downPulse[i] = max(downPulse[i - 1] + sequence[i], sequence[i]);
			upPulse[i] = max(upPulse[i - 1] - sequence[i], -sequence[i]);
		}
		else {
			downPulse[i] = max(downPulse[i - 1] - sequence[i], -sequence[i]);
			upPulse[i] = max(upPulse[i - 1] + sequence[i], sequence[i]);
		}
		answer = max3(upPulse[i], downPulse[i], answer);
	}
	return answer;
}
int main(void) {
	vector<int> v = { 2, 3, -6, 1, 3, -1, 2, 4 };
	solution(v);
}
//https://school.programmers.co.kr/learn/courses/30/lessons/161988

File: Algorithm/Programers_연속_펄스_부분_수열의_합.java
import java.util.*;
class Solution {
    public long solution(int[] sequence) {
        long[] sum = new long[sequence.length];
        long pre = 0, pulse = 1;
        long max = 0, min = 0;
        for (int i = 0; i < sequence.length; i++) {
            int seq = sequence[i];
            sum[i] = pre + pulse * seq;
            max = Math.max(max, sum[i]);
            min = Math.min(min, sum[i]);
            pre = sum[i];
            pulse *= -1;
        }
        return max - min;
    }
}
class main{
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] arr = {2, 3, -6, 1, 3, -1, 2, 4};
        System.out.println("sol = " + sol.solution(arr));
    }
}
//https://school.programmers.co.kr/learn/courses/30/lessons/161988